File: Source/AlphaTab/Platform/Svg/FontSizes.cs
using System;
using AlphaTab.Collections;
using AlphaTab.Platform.Model;

namespace AlphaTab.Platform.Svg
{
    /// <summary>
    /// This public class stores text widths for several fonts and allows width calculation 
    /// </summary>
    partial class FontSizes
    {
        public static byte[] Georgia;
        public static byte[] Arial;

        internal static FastDictionary<string, byte[]> FontSizeLookupTables;

        private static void Init()
        {
            Georgia = new byte[] { 3, 4, 5, 7, 7, 9, 8, 2, 4, 4, 5, 7, 3, 4, 3, 5, 7, 5, 6, 6, 6, 6, 6, 6, 7, 6, 3, 3, 7, 7, 7, 5, 10, 7, 7, 7, 8, 7, 7, 8, 9, 4, 6, 8, 7, 10, 8, 8, 7, 8, 8, 6, 7, 8, 7, 11, 8, 7, 7, 4, 5, 4, 7, 7, 6, 6, 6, 5, 6, 5, 4, 6, 6, 3, 3, 6, 3, 10, 6, 6, 6, 6, 5, 5, 4, 6, 5, 8, 6, 5, 5, 5, 4, 5, 7, 6, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 3, 4, 6, 7, 6, 7, 4, 6, 6, 10, 6, 6, 7, 0, 10, 7, 5, 7, 6, 6, 6, 6, 6, 3, 6, 6, 6, 6, 12, 12, 12, 5, 7, 7, 7, 7, 7, 7, 11, 7, 7, 7, 7, 7, 4, 4, 4, 4, 8, 8, 8, 8, 8, 8, 8, 7, 8, 8, 8, 8, 8, 7, 7, 6, 6, 6, 6, 6, 6, 6, 8, 5, 5, 5, 5, 5, 3, 3, 3, 3, 6, 6, 6, 6, 6, 6, 6, 7, 6, 6, 6, 6, 6, 5, 6 };
            Arial = new byte[] { 3, 3, 4, 6, 6, 10, 7, 2, 4, 4, 4, 6, 3, 4, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 3, 3, 6, 6, 6, 6, 11, 7, 7, 8, 8, 7, 7, 9, 8, 3, 6, 7, 6, 9, 8, 9, 7, 9, 8, 7, 7, 8, 7, 10, 7, 7, 7, 3, 3, 3, 5, 6, 4, 6, 6, 6, 6, 6, 3, 6, 6, 2, 2, 6, 2, 9, 6, 6, 6, 6, 4, 6, 3, 6, 6, 8, 6, 6, 6, 4, 3, 4, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 4, 6, 6, 6, 6, 3, 6, 4, 8, 4, 6, 6, 0, 8, 6, 4, 6, 4, 4, 4, 6, 6, 4, 4, 4, 4, 6, 9, 9, 9, 7, 7, 7, 7, 7, 7, 7, 11, 8, 7, 7, 7, 7, 3, 3, 3, 3, 8, 8, 9, 9, 9, 9, 9, 6, 9, 8, 8, 8, 8, 7, 7, 7, 6, 6, 6, 6, 6, 6, 10, 6, 6, 6, 6, 6, 3, 3, 3, 3, 6, 6, 6, 6, 6, 6, 6, 6, 7, 6, 6, 6, 6, 6, 6 };

            FontSizeLookupTables = new FastDictionary<string, byte[]>();
            FontSizeLookupTables["Arial"] = Arial;
            FontSizeLookupTables["'Arial'"] = Arial;
            FontSizeLookupTables["\"Arial\""] = Arial;
            FontSizeLookupTables["Georgia"] = Georgia;
            FontSizeLookupTables["'Georgia'"] = Georgia;
            FontSizeLookupTables["\"Georgia\""] = Georgia;
        }

        public const int ControlChars = 0x20;

        public static float MeasureString(string s, string family, float size, FontStyle style)
        {
            if (FontSizeLookupTables == null)
            {
                Init();
            }

            byte[] data;
            int dataSize = 11;
            if (FontSizeLookupTables.ContainsKey(family))
            {
                data = FontSizeLookupTables[family];
            }
            else
            {
                // TODO: on-the-fly-generation for all canvas types
                data = new byte[] { 8 };
            }

            float factor = 1;
            if ((style & FontStyle.Italic) != 0)
            {
                factor *= 1.2f;
            }
            if ((style & FontStyle.Bold) != 0)
            {
                factor *= 1.2f;
            }

            var stringSize = 0f;

            for (int i = 0; i < s.Length; i++)
            {
                var code = Math.Min(data.Length - 1, s[i] - ControlChars);
                if (code >= 0)
                {
                    stringSize += ((data[code] * size) / dataSize);
                }
            }

            return stringSize * factor;
        }
    }
}

File: packages/alphatab/src/platform/svg/FontSizes.ts
import { FontStyle, FontWeight } from '@src/model/Font';
import { Environment } from '@src/Environment';
import { MeasuredText } from '@src/platform/ICanvas';
import { WebPlatform } from '@src/platform/javascript/WebPlatform';

/**
 * Describes the sizes of a font for measuring purposes.
 * @internal
 */
export class FontSizeDefinition {
    /**
     * The widths of each character starting with the ascii code 0x20 at index 0.
     */
    public characterWidths: Uint8Array;

    /**
     * The heights of each character starting with the ascii code 0x20 at index 0.
     */
    public characterHeights: Uint8Array;

    public constructor(characterWidths: Uint8Array, characterHeights: Uint8Array) {
        this.characterWidths = characterWidths;
        this.characterHeights = characterHeights;
    }
}

/**
 * This public class stores text widths for several fonts and allows width calculation
 * @partial
 * @internal
 */
export class FontSizes {
    public static readonly fontSizeLookupTables: Map<string, FontSizeDefinition> = new Map<string, FontSizeDefinition>();

    public static readonly ControlChars: number = 0x20;

    /**
     * @target web
     * @partial
     */
    public static generateFontLookup(family: string): void {
        if (FontSizes.fontSizeLookupTables.has(family)) {
            return;
        }

        if (!Environment.isRunningInWorker && Environment.webPlatform !== WebPlatform.NodeJs) {
            const canvas: HTMLCanvasElement = document.createElement('canvas');
            const measureContext: CanvasRenderingContext2D = canvas.getContext('2d')!;
            const measureSize = 11;
            measureContext.font = `${measureSize}px ${family}`;
            const widths: number[] = [];
            const heights: number[] = [];
            for (let i: number = FontSizes.ControlChars; i < 255; i++) {
                const s: string = String.fromCharCode(i);
                const metrics = measureContext.measureText(s);
                widths.push(metrics.width);

                const height = metrics.actualBoundingBoxDescent + metrics.actualBoundingBoxAscent;
                heights.push(height);
            }

            const data: FontSizeDefinition = new FontSizeDefinition(new Uint8Array(widths), new Uint8Array(heights));
            FontSizes.fontSizeLookupTables.set(family, data);
        } else {
            const data: FontSizeDefinition = new FontSizeDefinition(new Uint8Array([8]), new Uint8Array([10]));
            FontSizes.fontSizeLookupTables.set(family, data);
        }
    }

    public static measureString(
        s: string,
        families: string[],
        size: number,
        style: FontStyle,
        weight: FontWeight
    ): MeasuredText {
        let data: FontSizeDefinition;
        const dataSize: number = 11;
        let family = families[0]; // default to first font

        // find a font which is maybe registered already
        for (let i = 0; i < families.length; i++) {
            if (FontSizes.fontSizeLookupTables.has(families[i])) {
                family = families[i];
                break;
            }
        }

        if (!FontSizes.fontSizeLookupTables.has(family)) {
            FontSizes.generateFontLookup(family);
        }
        data = FontSizes.fontSizeLookupTables.get(family)!;
        let factor: number = 1;
        if (style === FontStyle.Italic) {
            factor *= 1.1;
        }
        if (weight === FontWeight.Bold) {
            factor *= 1.1;
        }

        let stringSize: number = 0;
        let stringHeight: number = 0;
        for (let i: number = 0; i < s.length; i++) {
            const code: number = Math.min(data.characterWidths.length - 1, s.charCodeAt(i) - FontSizes.ControlChars);
            if (code >= 0) {
                stringSize += (data.characterWidths[code] * size) / dataSize;
                stringHeight = Math.max(stringHeight, (data.characterHeights[code] * size) / dataSize)
            }
        }

        // add a small increase of size for spacing/kerning etc.
        // we really need to improve the width calculation, maybe by using offscreencanvas?
        factor *= 1.07;

        return new MeasuredText(stringSize * factor, stringHeight);
    }
}
